probe_src/python/probe_py/manual/analysis.py

```diff
@@ -199,6 +237,126 @@ def add_edges(edges:list[tuple[Node, Node]], label:EdgeLabels) -> None:
+            if op.task_type != TaskType.TASK_PTHREAD and op.task_type != TaskType.TASK_ISO_C_THREAD:
+                
+                processNode1 = ProcessNode(pid = pid, cmd=tuple(cmd_map[pid]))
+                processNode2 = ProcessNode(pid = op.task_id, cmd=tuple(cmd_map[op.task_id]))
+                dataflow_graph.add_node(processNode1, label = processNode1.cmd)
+                dataflow_graph.add_node(processNode2, label = processNode2.cmd)
+                dataflow_graph.add_edge(processNode1, processNode2)
+            target_nodes[op.task_id] = list()
+        elif isinstance(op, WaitOp) and op.options == 0:
+            for node in target_nodes[op.task_id]:
+                traverse_hb_for_dfgraph(process_tree_prov_log, node, traversed, dataflow_graph, file_version_map, shared_files, cmd_map)
+                traversed.add(node[2])
+        # return back to the WaitOp of the parent process
+        if isinstance(next_op, WaitOp):
+            if next_op.task_id == starting_pid or next_op.task_id == starting_op.pthread_id:
+                return
+    return
+
+def list_edges_from_start_node(graph: nx.DiGraph, start_node: Node) -> list[EdgeType]:
+    all_edges = list(graph.edges())
+    start_index = next(i for i, edge in enumerate(all_edges) if edge[0] == start_node)
+    ordered_edges = all_edges[start_index:] + all_edges[:start_index] 
+    return ordered_edges
+
+def provlog_to_dataflow_graph(process_tree_prov_log: ProvLog) -> nx.DiGraph:
+    try:
+        dataflow_graph = nx.DiGraph()
+        file_version_map = collections.defaultdict[InodeOnDevice, int](lambda: 0)
+        process_graph = provlog_to_digraph(process_tree_prov_log)
+        root_node = [n for n in process_graph.nodes() if process_graph.out_degree(n) > 0 and process_graph.in_degree(n) == 0][0]
+        traversed: set[int] = set()
+        cmd:list[str] = []
+        cmd_map = collections.defaultdict[int, list[str]](list)
+        for edge in list(nx.edges(process_graph))[::-1]:
+            pid, exec_epoch_no, tid, op_index = edge[0]
+            op = prov_log_get_node(process_tree_prov_log, pid, exec_epoch_no, tid, op_index).data
+            if isinstance(op, OpenOp):
+                file = op.path.path.decode("utf-8")
+                if file not in cmd and file!="/dev/tty": 
+                    cmd.insert(0, file)
+            elif isinstance(op, InitExecEpochOp):
+                cmd.insert(0, op.program_name.decode("utf-8"))
+                if pid == tid and exec_epoch_no == 0:
+                    cmd_map[tid] = cmd
+                    cmd = []
+        shared_files:set[InodeOnDevice] = set()
+        traverse_hb_for_dfgraph(process_tree_prov_log, root_node, traversed, dataflow_graph, file_version_map, shared_files, cmd_map)
+        pydot_graph = nx.drawing.nx_pydot.to_pydot(dataflow_graph)
+        dot_string = pydot_graph.to_string()
+        return dot_string
+    except Exception as e:
+        traceback.print_exc()
+        print(str(e))
```

Comment:
We should just let the exception bubble up to the top.

Reply:
Should I include the try..except block in `dataflow-graph` function of `cli.py` instead of this?